[a11y] Semanctis flag refactor step 4: web and updateNode

hannah-hyj opened the pull request:
issue: https://github.com/flutter/flutter/issues/166101, overall goal is to update semantics flag to be a struct/class to support more than 32 flags.

step 1: https://github.com/flutter/flutter/pull/167421 Update semantics_node.h and dart:ui
step 2: https://github.com/flutter/flutter/pull/167738 Update Embedder part to use a struct instead of a int bit mask.
step 3: https://github.com/flutter/flutter/pull/167771 Update Framework use the SemanticsFlags class instead of bitmask
step 4 (this PR) Update web engine to use the new class and update SemanticsUpdateBuilder.updateNode to pass a list of bools instead of bitmask 

TODO:

flutter_tester
use the SemanticsFlags class instead of bitmask

[apicheck_test.dart](https://github.com/flutter/flutter/pull/167421/files#diff-69aefaacf1041f639974044962123bfae0756ce86032ac1f26256099425d7a5a)
Add this test back

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

hannah-hyj committed:
updates

Update semantics.dart

updates

Update semantics_update_builder.cc

updateNode use a list of bools

updateNode use a list of bools

Update ui_test.dart

1

1

hannah-hyj committed:
Update semantics_test.dart

tests

Update semantics_test.dart

update tests

chunhtai commented on engine/src/flutter/lib/ui/semantics.dart:
```diff
@@ -1668,7 +1746,7 @@ abstract class SemanticsUpdateBuilder {
   void updateNode({
     required int id,
-    required int flags,
+    required List<bool> flags,
```
Can we use NativeFieldWrapperClass1 for SemanticsFlags so that we don't do this serialization?

See StringAttribute class for example

hannah-hyj committed:
extends NativeFieldWrapperClass1 ?

Update semantics.dart

extends NativeFieldWrapperClass1 ?

SemanticsFlags extends NativeFieldWrapperClass1

Revert " extends NativeFieldWrapperClass1 ?"

This reverts commit df737e1ca14b6e8350b191f82db7c7097532b8e8.

extends NativeFieldWrapperClass1

extends NativeFieldWrapperClass1

lint

 resolve comments

Update semantics_flags.h

resolve comments

lint

fix tests

test

Revert "Update README.md"

This reverts commit 016aa67827dcc9a1616d609c738f2e46241b7d88.

Update README.md

Update semantics_test.dart

test

update tests

Update semantics_flags.cc

lint

chunhtai commented on engine/src/flutter/lib/ui/semantics.dart:
```diff
@@ -1062,10 +1062,116 @@ class SemanticsFlags {
 
   /// The set of semantics flags with every flag set to false.
-  static const SemanticsFlags kNone = SemanticsFlags();
+  static SemanticsFlags kNone = SemanticsFlags();
```
```suggestion
  static SemanticsFlags none = SemanticsFlags();
```

chunhtai commented on engine/src/flutter/lib/ui/semantics/semantics_flags.cc:
```diff
@@ -0,0 +1,67 @@
+  native_semantics_flags->AssociateWithDartWrapper(semantics_flags_handle);
+
+  native_semantics_flags->flags_ =
+      SemanticsFlags{.hasCheckedState = hasCheckedState};
```
what about other parameters?

chunhtai commented on engine/src/flutter/lib/web_ui/lib/semantics.dart:
```diff
@@ -303,6 +303,44 @@ class SemanticsFlags {
 
   static const SemanticsFlags kNone = SemanticsFlags();
 
+  factory SemanticsFlags.fromBoolList(List<bool> flags) {
```
this is not used right?

chunhtai commented on engine/src/flutter/lib/ui/semantics/semantics_flags.h:
```diff
@@ -0,0 +1,109 @@
+
+namespace flutter {
+
+struct SemanticsFlags;
```
unnecessary forward declaration.

chunhtai commented on engine/src/flutter/lib/ui/semantics/semantics_flags.h:
```diff
@@ -0,0 +1,109 @@
+
+struct SemanticsFlags;
+
+using SemanticsFlagsPtr = std::shared_ptr<flutter::SemanticsFlags>;
```
what are these for?

hannah-hyj replied:
copy paste mistakes! 😅

hannah-hyj committed:
fix FFI issue

chunhtai reviewed (approved):
LGTM
